### PR TITLE
Online SSIM and MS-SSIM Computation

### PR DESCRIPTION
**Pull request by Queuecumber**
## What does this PR do?

Working on #1224 
Fixes https://github.com/Lightning-AI/metrics/issues/1213

This PR fixes SSIM and MS-SSIM by allowing them to compute using running stats. This will make them actually usable in real situations.

Also I fixed several bugs that I found in these modules and changed the defaults to match pytorch-msssim

All tests should pass but to be honest I'm not sure how good those tests were anyway

I verified that this matches the known-good pytorch-mssim library to ~6 decimal places with default settings which was not the case before and I will be using this code in production in the coming week to see how it does.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Queuecumber** [comment]
I'm not sure what the best way to fix the mypy failure is, would appreciate some guidance on that. Explicitly annotating the variable as a `Tensor` didn't do it

**Comment by Queuecumber** [comment]
@SkafteNicki I could use some eyes on this, I've been using it in production and it seems to work well

File: src/torchmetrics/image/ssim.py [hunk 1]
```diff
@@ -55,9 +55,9 @@ class StructuralSimilarityIndexMeasure(Metric):
     Example:
         >>> from torchmetrics import StructuralSimilarityIndexMeasure
         >>> import torch
-        >>> preds = torch.rand([16, 1, 16, 16])
+        >>> preds = torch.rand([3, 3, 256, 256])
         >>> target = preds * 0.75
-        >>> ssim = StructuralSimilarityIndexMeasure()
+        >>> ssim = StructuralSimilarityIndexMeasure(data_range=1.0)
```
Comment (justusschock) [inline src/torchmetrics/image/ssim.py:60]: why can't we stick to `None` here?

File: src/torchmetrics/functional/image/ssim.py [hunk 2]
```diff
@@ -468,10 +462,10 @@ def multiscale_structural_similarity_index_measure(
 
     Example:
         >>> from torchmetrics.functional import multiscale_structural_similarity_index_measure
-        >>> preds = torch.rand([1, 1, 256, 256], generator=torch.manual_seed(42))
+        >>> preds = torch.rand([3, 3, 256, 256], generator=torch.manual_seed(42))
         >>> target = preds * 0.75
-        >>> multiscale_structural_similarity_index_measure(preds, target)
-        tensor(0.9558)
+        >>> multiscale_structural_similarity_index_measure(preds, target, data_range=1.0)
```
Comment (justusschock) [inline src/torchmetrics/functional/image/ssim.py:467]: why can't we stick to `None` here? that seems to be the most intuitive one for me.

**Review by justusschock — COMMENTED** [review]
It seems strange to me, that `compute` is called in `update` (this kind of breaks our paradigm), although the implementation seems to be correct.

**Comment by justusschock** [comment]
Could you also add tests against pytorch-mssim as this seems to also support batching (which is lacking in sklearn)?

File: src/torchmetrics/image/ssim.py (hunk 1 above)
Comment (Queuecumber) [reply src/torchmetrics/image/ssim.py:60]: I guess there's no reason why we couldn't since this is just a test case that we want to come out the same every time. Technically the range for `rand` is [0, 1) though

File: src/torchmetrics/functional/image/ssim.py (hunk 2 above)
Comment (Queuecumber) [reply src/torchmetrics/functional/image/ssim.py:467]: Actually I think `None` is the least intuitive since it estimates the min and max from the data. 

If this was real data (image for example) you know what the theoretical lower bound and upper bound on the values are and you wouldnt want to estimate them from the data. Otherwise you might get different ranges over different batches and the SSIM values wouldn't be directly comparable.

It's counterintuitive because who would want that?

**Comment by Queuecumber** [comment]
Yeah I will add tests against pytorch_msssim, do you want to leave the sklearn tests in place?

BTW we found an interesting issue with pytorch_msssim while testing this PR internally. It really doesn't play nice with fp16 AMP. Torchmetrics handles it much better. Irrelevant for the test cases but thought I would point it out

**Comment by Queuecumber** [comment]
> It seems strange to me, that `compute` is called in `update` (this kind of breaks our paradigm), although the implementation seems to be correct.

There should be a way to refactor it so that it looks more like the PSNR metric which does an `_update` call which is computing sum of squared error and then compute is more or less just reducing it. 

I'm fine implementing that refactor if you want

**Comment by Queuecumber** [comment]
Just following up, my plan is to do the following 

1. Rename `_ssim_update` to `_ssim_check_inputs` (open to suggestions on that name)
2. Rename `_ssim_compute` to `_ssim_update` and remove the reduction logic (but unsure of where to put the sum)
3. Add `_ssim_compute` which does the reduction (a little weird since in this case it's literally just calling reduce)
4. Update `structural_similarity_index_measure` to call these new functions correctly

**Comment by justusschock** [comment]
> do you want to leave the sklearn tests in place?

Ideally yes, as this is the most trusted lib for non-batched stuff. Maybe we can then reduce the tests compared to sklearn and add the more thorough testing with the other lib, but some sklearn tests should stay.


> Just following up, my plan is to do the following
>
>Rename _ssim_update to _ssim_check_inputs (open to suggestions on that name)
Rename _ssim_compute to _ssim_update and remove the reduction logic (but unsure of where to put the sum)
Add _ssim_compute which does the reduction (a little weird since in this case it's literally just calling reduce)
Update structural_similarity_index_measure to call these new functions correctly


Sounds good to me!

**Comment by Queuecumber** [comment]
OK I did the refactoring we discussed and added test cases for SSIM again pytorch-msssim (the MS-SSIM tests were already using it)

File: requirements/test.txt [hunk 3]
```diff
@@ -14,3 +14,4 @@ fire
 
 cloudpickle>=1.3
 scikit-learn>1.0, <1.1.1
+pytorch-msssim==0.2.1
```
Comment (SkafteNicki) [inline requirements/test.txt:17]: update this file instead:
https://github.com/Lightning-AI/metrics/blob/master/requirements/image_test.txt
```suggestion
```

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Review by justusschock — APPROVED** [review]
Thanks for doing. this refactor! LGTM!
